ratecalculator.go

```diff
@@ -0,0 +1,50 @@
+package main
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+var rc *rateCalculator
+
+
+type rateCalculator struct {
+	rates     []int64
+	prev      int64
+	counter   *int64
+	startTime time.Time
+}
+
+func newRateCalculator(keepIntervals int, interval time.Duration, counter *int64) *rateCalculator {
+	r := &rateCalculator{
+		rates:     make([]int64, keepIntervals),
+		counter:   counter,
+		startTime: time.Now(),
+	}
+
+	go r.updateRates(interval)
+
+	return r
+}
+
+func (r *rateCalculator) updateRates(interval time.Duration) {
+	for {
+		now := time.Now()
+		next := now.Truncate(interval).Add(interval)
+		time.Sleep(next.Sub(now))
+
+		cur := atomic.LoadInt64(r.counter)
+		rate := int64(float64(cur-r.prev) / interval.Seconds())
+		copy(r.rates[1:], r.rates)
```

Comment:
I know this isn't your code, but since you made me look at it again... This is racy and needs locking. `r.rates` is updated by one goroutine here and checked by another in `r.rate()` below. We need a `sync.Mutex` around the update and the read.